### ISSUE TITLE
Incompatibility between AppArmor and PodSecurityPolicy in MutatingWebhookAdmission

### ISSUE THREAD

**Issue by davidhao3300** [kubernetes/kubernetes#69953]
**Is this a BUG REPORT or FEATURE REQUEST?**: Bug Report
/kind bug

**What happened**:
PodSecurityPolicy is enabled, and allows everything, but has the `apparmor.security.beta.kubernetes.io/allowedProfileNames` and `apparmor.security.beta.kubernetes.io/defaultProfileName` annotations set.
Installed the Istio sidecar injector, and the injector attempts to inject a sidecar container. The error:
`replicaset-controller  Error creating: pods "{POD NAME}" is forbidden: unable to validate against any pod security policy: []`. Removing the apparmor annotations allow the sidecar to be created.
See the Additional Information for my guess as to what's happening based on reading the source code.

**What you expected to happen**:
The PodSecurityPolicy accepting the annotation change, or having a field in the securityContext to set the profileName for the sidecar.

**How to reproduce it (as minimally and precisely as possible)**:
We use a pretty modified version of istio to not have to include a privileged init container and enforce a more restricted PodSecurityPolicy. The below should reproduce with a vanilla version of Istio.

1. Install Istio sidecar injector, or anything that uses the MutatingAdmissionWebhook to add a container a pod that's being created.
2.  Create a PodSecurityPolicy that allows anything, but has the `apparmor.security.beta.kubernetes.io/allowedProfileNames` and `apparmor.security.beta.kubernetes.io/defaultProfileName` annotations set.
3. Observe the error described above.

**Anything else we need to know?**:
Looking through the source code, my guess is that the following is happening:
1. The injector calls PATCH on the pod to add the new container, resulting in the code execution eventually reaching https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L162
2. https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L261-L268 asks the PodSecurityPolicy to show what the pod has to look like in order to be accepted.
3. AppArmor says to add the annotation `apparmor.security.beta.kubernetes.io/...` because it's unset, and we've specified a default value in the PodSecurityPolicy. https://github.com/kubernetes/kubernetes/blob/master/pkg/security/podsecuritypolicy/provider.go#L182 sets the annotation in the pod. This means that step 2 rejects the PodSecurityPolicy because the returned pod is different.
4. Because no PodSecurityPolicy returns the exact pod copy, we exit the loop and return nil.
5. The next line in admission.go compares the returned nil against the original pod and returns the vague error message.

I believe another fix is to make the sidecar injector provide that annotation, but that seems less ideal than addressing this issue in some way in this codebase.

**Environment**:
- Kubernetes version (use `kubectl version`): `Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.8", GitCommit:"7eab6a49736cc7b01869a15f9f05dc5b49efb9fc", GitTreeState:"clean", BuildDate:"2018-09-14T15:54:20Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}`. Based on the source code, I believe this extends to the more recent versions as well.
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by davidhao3300** [comment]
Based on the OWNERS files in pkg/admission/security and pkg/security/podsecuritypolicy , I think this is for @kubernetes/sig-api-machinery-bugs and @kubernetes/sig-auth-bugs

**Comment by liggitt** [comment]
anything modifying pods via a mutating admission webhook must produce a pod that the validation phase of admission will accept as-is. this sounds like a similar ordering issue as described in https://github.com/kubernetes/kubernetes/issues/64333

One possibile scenario:
* there are multiple PSPs your pod could be admitted under
* one of the PSPs does not specify a default/allowed apparmor profile, and allows the pod as-is (prior to your istio-contributed container)
* your pod passes the in-tree mutating admission phase without getting the default apparmor annotation added
* then, your webhook adds a container which changes the makeup of the pod, and changes which PSPs accept the pod
* in the validation phase, that narrower set of policies match, but all of the policies that would admit the pod require changes (like adding apparmor annotations). because they are in the validation phase, those changes cannot be made, and they must reject the pod instead

some more information would be helpful in determining if this is what is happening:
* do you have multiple pod security policies in use, available to these pods? can you provide them? do any of them omit apparmor allowedProfileNames/defaultProfileName annotations?
* can you provide the original replicaset, and the container your version of istio is attempting to add?

**Comment by davidhao3300** [comment]
Yeah, I saw the other issues caused by the fix ordering introduced by the 1.9->1.10 upgrade (we were on 1.9 before, and this was working then).

The information requested:
We only bind one PodSecurityPolicy to `system:serviceaccounts`. The contents of the PodSecurityPolicy is as follows:
```
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  name: restricted
  annotations:
    seccomp.security.alpha.kubernetes.io/allowedProfileNames: docker/default
    apparmor.security.beta.kubernetes.io/allowedProfileNames: runtime/default
    seccomp.security.alpha.kubernetes.io/defaultProfileName: docker/default
    apparmor.security.beta.kubernetes.io/defaultProfileName: runtime/default
spec:
  privileged: false
  allowPrivilegeEscalation: false
  allowedCapabilities: []
  volumes:
    - configMap
    - emptyDir
    - projected
    - secret
    - downwardAPI
    - persistentVolumeClaim
    - nfs
  hostNetwork: false
  hostIPC: false
  hostPID: false
  runAsUser:
    rule: RunAsAny
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  fsGroup:
    rule: RunAsAny
  readOnlyRootFilesystem: false
```
I've made sure that the error was from the apparmor annotation only - relaxing/removing the other restrictions has no effect.

We do have 2 other, more relaxed PodSecurities, but one of them is bound to `system:serviceaccounts:kube-system` and `system:nodes`, and the other (called kube-system, doesn't seem to have been created manually by us) doesn't seem to be bound to anything. Neither of them have the apparmor annotation set.

I think the last 2 steps are basically what's happening, since the AppArmor annotation has to be set per-CONTAINER, not per-POD, as in there'll be an annotation `container.apparmor.security.beta.kubernetes.io/{main container name}: {default value}` when the pod is first created, and then once the sidecar is added, the PSP tries to add `container.apparmor.security.beta.kubernetes.io/{sidecar container name}: {default value}`, so the pod gets rejected.

**Comment by davidhao3300** [comment]
this is a shortened version of the original pod that's created if i remove the AppArmor annotations:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/psp: restricted
    seccomp.security.alpha.kubernetes.io/pod: docker/default
    sidecar.istio.io/status: '{"version":"e4509f7d20f041e639ca187af6389ac9fe04290f3910a2dafba02a2bcd3a39e5","initContainers":null,"containers":["istio-proxy"],"volumes":["istio-envoy","istio-certs"],"imagePullSecrets":null}'
  name: {pod name}
  namespace: {namespace}
spec:
  containers:
  - {main container spec}
  - args:
    - proxy
    - sidecar
    - --configPath
    - /etc/istio/proxy
    - --binaryPath
    - /usr/local/bin/envoy
    - --serviceCluster
    - kafka
    - --drainDuration
    - 45s
    - --parentShutdownDuration
    - 1m0s
    - --discoveryAddress
    - istio-pilot.istio-system:15007
    - --discoveryRefreshDelay
    - 1s
    - --zipkinAddress
    - zipkin.istio-system:9411
    - --connectTimeout
    - 10s
    - --statsdUdpAddress
    - istio-statsd-prom-bridge.istio-system:9125
    - --proxyAdminPort
    - "15000"
    - --controlPlaneAuthPolicy
    - NONE
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    - name: ISTIO_META_POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: ISTIO_META_INTERCEPTION_MODE
      value: REDIRECT
    image: docker.io/istio/proxyv2:1.0.2
    imagePullPolicy: IfNotPresent
    name: istio-proxy
    resources:
      limits:
        cpu: 10m
        memory: 512Mi
      requests:
        cpu: 10m
        memory: 512Mi
    securityContext:
      readOnlyRootFilesystem: true
      runAsUser: 1337
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/istio/proxy
      name: istio-envoy
    - mountPath: /etc/certs/
      name: istio-certs
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: ip-10-62-51-44.us-west-2.compute.internal
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 0
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - configMap:
      defaultMode: 420
      name: kafka
    name: encoded-keys
  - name: default-token-hrtn4
    secret:
      defaultMode: 420
      secretName: default-token-hrtn4
  - emptyDir:
      medium: Memory
    name: istio-envoy
  - name: istio-certs
    secret:
      defaultMode: 420
      optional: true
      secretName: istio.default
```

**Comment by liggitt** [comment]
@tallclair do you know of plans to graduate apparmor support to fields, and whether a pod-wide default was considered?

**Comment by davidhao3300** [comment]
Agreed, making the Istio sidecar injector support custom annotations is an option, but I think this is something that should be addressed eventually.
And yeah, we wanted to make the controllers use those PSPs on purpose.

**Comment by liggitt** [comment]
I'm going to close this as a duplicate of #64333, and will track it there, since the issue is the same (per-container fields required by admission getting injected after in-tree admission has a chance to make changes automatically)

/close

**Comment by davidhao3300** [comment]
okay, sounds good

**Comment by davidhao3300** [comment]
Would be great to have a pod-wide default if possible. In the mean time, I'm opening an issue with istio to customize annotations of the sidecar container as a workaround

**Comment by tallclair** [comment]
Yeah, agreed. I think that is blocked on migrating the annotations to fields, which we should prioritize soon.

**Comment by liggitt** [comment]
xref https://github.com/kubernetes/kubernetes/pull/78080/
